src/main/java/uk/nhs/hee/tis/trainee/credentials/service/IssuanceService.java

```diff
@@ -78,4 +94,67 @@ public Optional<URI> startCredentialIssuance(String authToken, CredentialDto dto
 
     return gatewayService.getCredentialUri(dto, nonce.toString(), internalState.toString());
   }
+
+  /**
+   * Complete the credential issuance process.
+   *
+   * @param code             The code provided by the credential gateway.
+   * @param state            The state set in the initial gateway request.
+   * @param error            The error text, if the credential was not issued.
+   * @param errorDescription The error description, if the credential was not issued.
+   * @return The built redirect URI for completed issuance.
+   */
+  public URI completeCredentialVerification(String code, String state, String error,
+      String errorDescription) {
+    UUID stateUuid = UUID.fromString(state);
+    CredentialMetadata credentialMetadata = null;
+
+    if (error == null && code != null) {
+      log.info("Credential was issued successfully.");
+      URI tokenEndpoint = URI.create(properties.tokenEndpoint());
+      URI redirectEndpoint = URI.create(properties.redirectUri());
+      Claims claims = gatewayService.getTokenClaims(tokenEndpoint, redirectEndpoint, code, null);
+      UUID nonceUuid = UUID.fromString(claims.get("nonce", String.class));
+
+      Optional<String> traineeId = cachingDelegate.getTraineeIdentifier(stateUuid);
+      Optional<CredentialDto> credentialData = cachingDelegate.getCredentialData(nonceUuid);
+      if (traineeId.isPresent() && credentialData.isPresent()) {
+        IssueResponseDto issueResponseDto = fromIssuedResponse(claims, traineeId.get());
+        credentialMetadata = credentialMetadataMapper
+            .toCredentialMetadata(traineeId.get(), credentialData.get(), issueResponseDto);
+      }
+    } else {
+      log.info("Credential was not issued.");
+    }
+
+    if (credentialMetadata != null) {
+      credentialMetadataRepository.save(credentialMetadata);
+    }
+
+    Optional<String> clientState = cachingDelegate.getClientState(stateUuid);
+
+    // Build and return the redirect_uri
+    return UriComponentsBuilder.fromUriString("/credential-issued")
```

Comment:
I guess there's no need to return the code as well at this point

Reply:
I've done the same thing for verification, though there are two redirects there (either `credential-verified` or `invalid-credential`). It feels like we **could** move them to the app config files, but they would still be hard coded rather than and env var because it doesn't feel like something we'd change per deployment.